File: 总题库/272.ClosestBinarySearchTreeValue2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def closestKValues(self, root: TreeNode, target: float, k: int) -> List[int]:
        # 在中序遍历的过程中维护result数组；
        if not root: return []
        stack = []
        result = []
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            if len(result) < k: 
                result.append(root.val)
            elif abs(result[0] - target) > abs(root.val - target):
                result.pop(0)
                result.append(root.val)
            else: 
                # 因为中序遍历得到的是一个升序序列，此时可以进行剪枝；
                break
            root = root.right
        return result

File: 总题库/272.ClosestBinarySearchTreeValue2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// maintain the result k values when doing inorder traveral
class Solution {
    vector<int> inorderTraversal;
    queue<int> q;

    void inorder(TreeNode* root, double target, int k) {
        stack<TreeNode*> s;
        TreeNode* temp = root;
        while (temp != NULL || !s.empty()) {
            while (temp != NULL) {
                s.push(temp);
                temp = temp->left;
            }
            temp = s.top();
            s.pop();
            if (q.size() < k) {
                q.push(temp->val);
            }
            else if (abs(temp->val - target) < abs(q.front() - target)) {
                q.pop();
                q.push(temp->val);
            }
            else {
                break;
            }
            temp = temp->right;
        }
    }

public:
    vector<int> closestKValues(TreeNode* root, double target, int k) {
        vector<int> res;
        inorder(root, target, k);
        while (!q.empty()) {
            res.push_back(q.front());
            q.pop();
        }

        return res;
    }
};